File: bitdrift-shop/kotlin-multiplatform/iosApp/iosApp/ScreenLogger.swift
import Foundation
import Capture
import os.log

// Centralised logging that writes to both os.log (local console) and the
// bitdrift dashboard.  All screen transitions and key events flow through here.
enum ScreenLogger {
    private static let osLogger = os.Logger(
        subsystem: "com.example.shoppingdemo.kmp",
        category: "ScreenLogger"
    )

    // Called by ScreenContainer.onAppear for every screen transition.
    static func logScreenView(_ screenName: String) {
        osLogger.info("_screen_name: \(screenName)")
        Logger.logScreenView(screenName: screenName)
    }

    static func logInfo(_ message: String, fields: [String: String] = [:]) {
        printLog("INFO", message, fields)
        Logger.logInfo(message, fields: fields)
    }

    static func logWarning(_ message: String, fields: [String: String] = [:]) {
        printLog("WARNING", message, fields)
        Logger.logWarning(message, fields: fields)
    }

    static func logError(_ message: String, fields: [String: String] = [:]) {
        printLog("ERROR", message, fields)
        Logger.logError(message, fields: fields)
    }

    static func logError(_ message: String, error: Error, fields: [String: String] = [:]) {
        printLog("ERROR", message, fields)
        Logger.logError(message, error: error, fields: fields)
    }

    // MARK: - Private

    private static func printLog(_ level: String, _ message: String, _ fields: [String: String]) {
        var output = "[\(level)] \(message)"
        if !fields.isEmpty {
            let sorted = fields.sorted { $0.key < $1.key }
                .map { "\($0.key)=\($0.value)" }
                .joined(separator: " | ")
            output += " | \(sorted)"
        }
        osLogger.info("\(output)")
    }
}

File: misc-demos/shoppingdemo-oteldemo/android/sdk/app/src/main/java/com/example/shoppingdemo/ScreenLogger.kt
package com.example.shoppingdemo

import android.util.Log
import io.bitdrift.capture.Capture.Logger

/**
 * Centralized logging for screen views and user actions.
 */
object ScreenLogger {

    private const val TAG = "ScreenLogger"

    fun logScreenView(screenName: String) {
        Log.d(TAG, "_screen_name: $screenName")
        // Workshop 2 (Screen Views): log each screen transition to bitdrift so it
        // appears as a breadcrumb in user journeys and powers Sankey diagrams in the dashboard.
        // https://docs.bitdrift.io/sdk/quickstart#android
        //
        // ScreenContainer already calls ScreenLogger.logScreenView via DisposableEffect for
        // every screen — it was previously only using Log.d. Wiring in Logger.logScreenView()
        // here is the single change that covers all 15 screens automatically.
        Logger.logScreenView(screenName)
    }

    fun logInfo(message: String, fields: Map<String, String> = emptyMap()) {
        printLog("INFO", message, fields)
        Logger.logInfo(fields) { message }
    }

    fun logError(message: String, fields: Map<String, String> = emptyMap()) {
        printLog("ERROR", message, fields)
        Logger.logError(fields) { message }
    }

    fun logSimulationStart(runs: Int) {
        logInfo("simulation_start", mapOf("total_runs" to runs.toString()))
    }

    fun logSimulationEnd(runs: Int) {
        logInfo("simulation_end", mapOf("total_runs" to runs.toString()))
    }

    private fun printLog(level: String, message: String, fields: Map<String, String>) {
        val output = buildString {
            append("[$level] $message")
            if (fields.isNotEmpty()) {
                append(" | ")
                append(fields.entries.sortedBy { it.key }.joinToString(" | ") { "${it.key}=${it.value}" })
            }
        }
        Log.d(TAG, output)
    }
}
